File: apps/dashboard/.claude.backup.2026-05-14T00-00-02/scripts/qa-run.js
#!/usr/bin/env bun
'use strict';
/**
 * QA-RUN: Executes Acceptance Criteria defined in a spec file.
 *
 * Reads `.claude/specs/{spec}.md` (or `.claude/spec/active/{spec}/spec.md`),
 * extracts the "Acceptance Criteria" section, runs each AC command,
 * and emits a `qa.result` event to the harness event log.
 *
 * Wave 10 — Dev/QA contract enforcement.
 *
 * Usage:
 *   bun .claude/scripts/qa-run.js --spec auth-login
 *   bun .claude/scripts/qa-run.js --spec auth-login --json
 *
 * Exported API:
 *   module.exports = { runQA };
 *   runQA({ spec, cwd }) → Promise<QAResult>
 *
 * @version 1.0.0
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const { emit } = require('../hooks/_lib/harness-event.js');

const AC_TIMEOUT_MS = 120_000; // 2 min per AC

// ── AC Parsing ─────────────────────────────────────────────────────────────────

/**
 * Extract the "Acceptance Criteria" section content from spec markdown.
 * Returns the raw section text (everything after the heading until the next ##-level heading),
 * or null if the section is not found.
 */
function extractACSection(markdown) {
  // Find "## Acceptance Criteria" heading (case-insensitive)
  const headingIdx = markdown.search(/^##\s+Acceptance\s+Criteria\s*$/im);
  if (headingIdx < 0) return null;

  const fromHere = markdown.slice(headingIdx);
  // Skip past the heading line
  const newlineIdx = fromHere.indexOf('\n');
  if (newlineIdx < 0) return '';
  const rest = fromHere.slice(newlineIdx + 1);

  // Terminate at the next ## heading
  const nextSection = rest.search(/^##\s/im);
  return nextSection >= 0 ? rest.slice(0, nextSection) : rest;
}

/**
 * Parse AC items from the section text.
 * Accepted formats:
 *   - [ ] AC-1: description — Command: `cmd`
 *   - [ ] AC-1: description — Command: cmd (no backticks)
 *   - [x] AC-1: (already checked — still executed)
 *
 * Returns array of { id, description, command }
 */
function parseACItems(sectionText) {
  if (!sectionText) return [];
  const items = [];
  const lines = sectionText.split('\n');

  for (const line of lines) {
    // Match: - [ ] AC-N: description — Command: `cmd` OR - [ ] AC-N: description — Command: cmd
    const m = line.match(
      /^\s*-\s*\[[ xX]\]\s*(AC-\d+)\s*:\s*(.+?)\s*(?:—|-{1,2})\s*Command\s*:\s*`?([^`\n]+)`?\s*$/i
    );
    if (!m) continue;
    const [, id, description, command] = m;
    items.push({ id: id.toUpperCase(), description: description.trim(), command: command.trim() });
  }
  return items;
}

// ── Spec Location ──────────────────────────────────────────────────────────────

/**
 * Find the spec file for a given spec name. Tries multiple locations:
 *   1. .claude/specs/{spec}.md
 *   2. .claude/spec/active/{spec}/spec.md
 *   3. .claude/spec/completed/{spec}/spec.md
 */
function findSpecFile(cwd, spec) {
  const candidates = [
    path.join(cwd, '.claude', 'specs', spec + '.md'),
    path.join(cwd, '.claude', 'spec', 'active', spec, 'spec.md'),
    path.join(cwd, '.claude', 'spec', 'completed', spec, 'spec.md'),
  ];
  for (const c of candidates) {
    if (fs.existsSync(c)) return c;
  }
  return null;
}

// ── AC Execution ───────────────────────────────────────────────────────────────

/**
 * Run a single AC command.
 * Returns { status: 'pass'|'fail'|'skip', exit, duration_ms, stderr_excerpt }
 */
function runACCommand(command, cwd) {
  const t0 = Date.now();

  try {
    execSync(command, {
      cwd,
      timeout: AC_TIMEOUT_MS,
      stdio: 'pipe',
      encoding: 'utf8',
      shell: true,
      windowsHide: true,
    });
    const duration_ms = Date.now() - t0;
    return { status: 'pass', exit: 0, duration_ms, stderr_excerpt: '' };
  } catch (err) {
    const duration_ms = Date.now() - t0;

    // Command not found or spawn error (ENOENT = shell not found, very unusual) — skip
    if (err.code === 'ENOENT') {
      return { status: 'skip', exit: null, duration_ms, stderr_excerpt: 'command not found' };
    }

    // Timeout — skip (env error, not a real AC failure)
    if (err.killed || err.code === 'ETIMEDOUT') {
      return {
        status: 'skip',
        exit: null,
        duration_ms,
        stderr_excerpt: `timeout after ${AC_TIMEOUT_MS}ms`,
      };
    }

    // Real failure (non-zero exit code)
    const stderrRaw = (err.stderr || '').trim();
    const stdoutRaw = (err.stdout || '').trim();
    const combined = [stderrRaw, stdoutRaw].filter(Boolean).join(' ').slice(0, 100);
    return {
      status: 'fail',
      exit: typeof err.status === 'number' ? err.status : 1,
      duration_ms,
      stderr_excerpt: combined,
    };
  }
}

// ── Report Writing ─────────────────────────────────────────────────────────────

function writeQAReport(cwd, spec, payload) {
  try {
    const reportsDir = path.join(cwd, '.claude', '.qa-reports');
    fs.mkdirSync(reportsDir, { recursive: true });
    const reportPath = path.join(reportsDir, spec + '.json');
    fs.writeFileSync(reportPath, JSON.stringify(payload, null, 2), 'utf8');
  } catch (_) {
    // fail-open: report write failure does not affect qa result
  }
}

// ── Main runQA ─────────────────────────────────────────────────────────────────

/**
 * Run QA for a spec.
 *
 * @param {{ spec: string, cwd?: string }} opts
 * @returns {Promise<QAResult>}
 *
 * QAResult:
 * {
 *   spec: string,
 *   overall: 'pass' | 'fail' | 'skip',
 *   criteria: Array<{ id, status, exit, duration_ms, stderr_excerpt }>,
 *   skippedReason?: string,   // set when no AC section or no AC items
 *   report: string            // human-readable markdown
 * }
 */
async function runQA({ spec, cwd: cwdArg } = {}) {
  const cwd = cwdArg || process.cwd();

  // ── Locate spec ──────────────────────────────────────────────────────────────
  const specFile = findSpecFile(cwd, spec);
  if (!specFile) {
    const result = {
      spec,
      overall: 'skip',
      criteria: [],
      skippedReason: `Spec file not found for "${spec}" — tried .claude/specs/${spec}.md, .claude/spec/active/${spec}/spec.md`,
      report: `## QA Report for spec: ${spec}\n\n**SKIP** — spec file not found.\n`,
    };
    process.stderr.write(`[qa-run] ${result.skippedReason}\n`);
    return result;
  }

  // ── Read spec ────────────────────────────────────────────────────────────────
  let markdown;
  try {
    markdown = fs.readFileSync(specFile, 'utf8');
  } catch (err) {
    const result = {
      spec,
      overall: 'skip',
      criteria: [],
      skippedReason: `Cannot read spec file: ${err.message}`,
      report: `## QA Report for spec: ${spec}\n\n**SKIP** — cannot read spec file.\n`,
    };
    process.stderr.write(`[qa-run] ${result.skippedReason}\n`);
    return result;
  }

  // ── Extract Acceptance Criteria ───────────────────────────────────────────────
  const acSection = extractACSection(markdown);
  if (!acSection) {
    const result = {
      spec,
      overall: 'skip',
      criteria: [],
      skippedReason: 'No "Acceptance Criteria" section found in spec',
      report: `## QA Report for spec: ${spec}\n\n**SKIP** — no Acceptance Criteria section in spec.\n`,
    };
    process.stderr.write(`[qa-run] WARN: ${result.skippedReason}\n`);
    return result;
  }

  const acItems = parseACItems(acSection);
  if (acItems.length === 0) {
    const result = {
      spec,
      overall: 'skip',
      criteria: [],
      skippedReason: 'Acceptance Criteria section found but no parseable AC items',
      report: `## QA Report for spec: ${spec}\n\n**SKIP** — no parseable AC items found.\n`,
    };
    process.stderr.write(`[qa-run] WARN: ${result.skippedReason}\n`);
    return result;
  }

  // ── Execute each AC ──────────────────────────────────────────────────────────
  const criteria = [];
  let failCount = 0;
  let skipCount = 0;

  for (const ac of acItems) {
    const { status, exit, duration_ms, stderr_excerpt } = runACCommand(ac.command, cwd);
    criteria.push({ id: ac.id, status, exit, duration_ms, stderr_excerpt });
    if (status === 'fail') failCount++;
    if (status === 'skip') skipCount++;
  }

  const overall = failCount > 0 ? 'fail' : (skipCount === acItems.length ? 'skip' : 'pass');

  // ── Build report ──────────────────────────────────────────────────────────────
  const lines = [`## QA Report for spec: ${spec}`, ''];
  for (const c of criteria) {
    const icon = c.status === 'pass' ? '✅' : c.status === 'fail' ? '❌' : '⏭️';
    const exitStr = c.exit !== null ? `exit ${c.exit}` : 'n/a';
    const durStr = `${(c.duration_ms / 1000).toFixed(1)}s`;
    let line = `- ${c.id}: ${icon} ${c.status.toUpperCase()} — ${exitStr} (${durStr})`;
    if (c.stderr_excerpt) line += ` — stderr: ${c.stderr_excerpt.slice(0, 50)}`;
    lines.push(line);
  }
  lines.push('');
  lines.push(`**Overall**: ${overall.toUpperCase()} (${failCount} of ${acItems.length} failed)`);
  const report = lines.join('\n');

  // ── Payload ───────────────────────────────────────────────────────────────────
  const payload = { spec, overall, criteria };

  // ── Emit harness event ────────────────────────────────────────────────────────
  try {
    emit('qa.result', payload, { cwd, actor: { kind: 'script', id: 'qa-run' } });
  } catch (_) {
    // fail-open: event emission does not affect QA result
  }

  // ── Write sidecar ─────────────────────────────────────────────────────────────
  writeQAReport(cwd, spec, payload);

  return { ...payload, report };
}

// ── CLI entrypoint ────────────────────────────────────────────────────────────

if (require.main === module) {
  const args = process.argv.slice(2);
  let spec = null;
  let jsonOnly = false;
  let cwdArg = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--spec' && args[i + 1]) { spec = args[++i]; }
    else if (args[i] === '--json') { jsonOnly = true; }
    else if (args[i] === '--cwd' && args[i + 1]) { cwdArg = args[++i]; }
  }

  if (!spec) {
    process.stderr.write('Usage: node qa-run.js --spec <name> [--json] [--cwd <path>]\n');
    process.exit(1);
  }

  runQA({ spec, cwd: cwdArg || process.cwd() }).then(result => {
    if (jsonOnly) {
      process.stdout.write(JSON.stringify({
        event: 'qa.result',
        payload: { spec: result.spec, overall: result.overall, criteria: result.criteria },
      }, null, 2) + '\n');
    } else {
      process.stdout.write(result.report + '\n');
      if (result.skippedReason) {
        process.stderr.write(`[qa-run] Skipped: ${result.skippedReason}\n`);
      }
    }
    process.exit(result.overall === 'fail' ? 1 : 0);
  }).catch(err => {
    process.stderr.write(`[qa-run] Fatal error: ${err.message}\n`);
    process.exit(0); // fail-open
  });
}

module.exports = { runQA };

File: packages/rt/src/run/qa_run.rs
//! `mustard-rt run qa-run` — a port of `scripts/qa-run.js`.
//!
//! Executes the Acceptance Criteria defined in a spec file: locates the spec,
//! extracts the `## Acceptance Criteria` section, runs each AC command, and
//! emits a `qa.result` harness event plus a `qa` metric.
//!
//! Port note: the JS version shelled to `_lib/harness-event.js` and
//! `_lib/metrics-emit.js`. This port emits the event through `mustard_core`'s
//! [`JsonlEventStore`] and the metric through `mustard_core::metrics`.
//!
//! Fail-open: a missing spec or no AC section degrades to an `overall: skip`
//! result and exit `0`; an AC failure exits `1` (the JS contract).
//!
//! `--format json` (default) prints the `{ event, payload }` JSON the pipeline
//! consumes. `--format html` additionally writes a standalone HTML report to
//! `.claude/.qa-reports/{spec}.html` and prints its path on stderr; JSON is
//! still emitted on stdout — HTML is an artifact, never a replacement.

use crate::report::{table, Report};
use crate::run::env::{project_dir, session_id};
use crate::util::now_iso8601;
use mustard_core::io::event_store::{EventSink, JsonlEventStore};
use mustard_core::metrics::{emit_metric, MetricLine};
use mustard_core::model::event::{Actor, ActorKind, HarnessEvent, SCHEMA_VERSION};
use serde_json::{json, Value};
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::Instant;

/// Per-AC timeout (2 min), matching `AC_TIMEOUT_MS` in `qa-run.js`.
const AC_TIMEOUT_SECS: u64 = 120;

/// A parsed AC item: `- [ ] AC-N: description — Command: `cmd``.
struct AcItem {
    id: String,
    command: String,
}

/// One AC execution outcome.
struct AcResult {
    id: String,
    status: String,
    exit: Option<i64>,
    duration_ms: u128,
    stderr_excerpt: String,
}

/// Locate the spec file: `.claude/specs/{spec}.md`, then
/// `.claude/spec/active/{spec}/spec.md`, then `completed/`.
fn find_spec_file(cwd: &Path, spec: &str) -> Option<PathBuf> {
    let candidates = [
        cwd.join(".claude").join("specs").join(format!("{spec}.md")),
        cwd.join(".claude").join("spec").join("active").join(spec).join("spec.md"),
        cwd.join(".claude").join("spec").join("completed").join(spec).join("spec.md"),
    ];
    candidates.into_iter().find(|c| c.exists())
}

/// Extract the `## Acceptance Criteria` section body (heading line stripped),
/// recognizing the EN and PT headings via [`crate::run::spec_sections`].
fn extract_ac_section(markdown: &str) -> Option<String> {
    let lines: Vec<&str> = markdown.split('\n').collect();
    let start = lines
        .iter()
        .position(|l| crate::run::spec_sections::is_heading(l, "acceptanceCriteria"))?;
    let mut end = lines.len();
    for (i, l) in lines.iter().enumerate().skip(start + 1) {
        if l.starts_with("## ") {
            end = i;
            break;
        }
    }
    // Body only — drop the heading line itself.
    Some(lines[start + 1..end].join("\n"))
}

/// Parse `- [ ] AC-N: description — Command: `cmd`` lines.
///
/// JS regex (case-insensitive):
/// `^\s*-\s*\[[ xX]\]\s*(AC-\d+)\s*:\s*(.+?)\s*(?:—|-{1,2})\s*Command\s*:\s*`?([^`\n]+)`?\s*$`
fn parse_ac_items(section: &str) -> Vec<AcItem> {
    let mut items = Vec::new();
    for line in section.split('\n') {
        if let Some(item) = parse_ac_line(line) {
            items.push(item);
        }
    }
    items
}

/// Parse one AC line with plain string scanning (no regex crate available).
fn parse_ac_line(line: &str) -> Option<AcItem> {
    let t = line.trim_start();
    let rest = t.strip_prefix('-')?.trim_start();
    // `[ ]`, `[x]`, `[X]`.
    let rest = rest.strip_prefix('[')?;
    let mark = rest.chars().next()?;
    if !matches!(mark, ' ' | 'x' | 'X') {
        return None;
    }
    let rest = rest[mark.len_utf8()..].strip_prefix(']')?.trim_start();
    // `AC-<digits>`.
    let lower = rest.to_lowercase();
    if !lower.starts_with("ac-") {
        return None;
    }
    let after_ac = &rest[3..];
    let digits_end = after_ac.find(|c: char| !c.is_ascii_digit()).unwrap_or(after_ac.len());
    if digits_end == 0 {
        return None;
    }
    let id = format!("AC-{}", &after_ac[..digits_end]);
    let after_id = after_ac[digits_end..].trim_start();
    let after_colon = after_id.strip_prefix(':')?;
    // Find a `Command:` segment after an em-dash / hyphen separator.
    let lower_seg = after_colon.to_lowercase();
    let cmd_idx = lower_seg.find("command")?;
    // The char just before `command` (after trimming `:` and ws) should be a
    // separator — the JS pattern requires `—` / `-` / `--` before `Command`.
    let cmd_tail = &after_colon[cmd_idx + "command".len()..];
    let cmd_tail = cmd_tail.trim_start().strip_prefix(':')?.trim();
    let command = cmd_tail.trim_matches('`').trim().to_string();
    if command.is_empty() {
        return None;
    }
    Some(AcItem { id: id.to_uppercase(), command })
}

/// Build the platform shell invocation for an AC `command` string.
///
/// On non-Windows: `sh -c <command>` — `std`'s normal arg passing is correct
/// for `sh`, which parses argv entries directly.
///
/// On Windows: `cmd.exe` does **not** parse its command line via the
/// `CommandLineToArgvW` rules that `std`'s `Command::arg` quoting assumes, so
/// passing a complex `command` (quotes, `()`, `|`, `&&`) through `arg` corrupts
/// it. Instead, append the command verbatim with `CommandExt::raw_arg` (a SAFE
/// API — no `unsafe` needed) and invoke `cmd /S /C "<command>"`: with `/S` and a
/// command line whose first and last chars are quotes, `cmd` strips exactly
/// that outer quote pair and runs the remainder literally.
#[cfg(windows)]
fn build_shell_command(command: &str) -> Command {
    use std::os::windows::process::CommandExt;
    let mut c = Command::new("cmd");
    // Single verbatim argument: `/S /C "<command>"`. One `raw_arg` call so the
    // whole `cmd` command line is exactly this — no `std` quoting, no extra
    // separators between the `/S /C` flags and the quoted payload.
    c.raw_arg(format!("/S /C \"{command}\""));
    c
}

/// See the `#[cfg(windows)]` variant for the rationale.
#[cfg(not(windows))]
fn build_shell_command(command: &str) -> Command {
    let mut c = Command::new("sh");
    c.arg("-c").arg(command);
    c
}

/// Run one AC command. Mirrors the JS classification: `pass` (exit 0), `fail`
/// (non-zero exit), `skip` (timeout or spawn failure).
fn run_ac_command(command: &str, cwd: &Path) -> AcResult {
    let t0 = Instant::now();
    // POSIX-style AC commands assume a shell; use the platform shell. Windows
    // AC are documented to be cross-shell-safe (`node -e`, `bash -c`).
    let mut cmd = build_shell_command(command);
    cmd.current_dir(cwd);

    // No native wait-with-timeout in std; spawn + poll.
    let child = cmd
        .stdin(std::process::Stdio::null())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn();
    let mut child = match child {
        Ok(c) => c,
        Err(_) => {
            return AcResult {
                id: String::new(),
                status: "skip".to_string(),
                exit: None,
                duration_ms: t0.elapsed().as_millis(),
                stderr_excerpt: "command not found".to_string(),
            };
        }
    };

    let deadline = Instant::now() + std::time::Duration::from_secs(AC_TIMEOUT_SECS);
    loop {
        match child.try_wait() {
            Ok(Some(status)) => {
                let out = child.wait_with_output().ok();
                let (stderr, stdout) = out
                    .map(|o| {
                        (
                            String::from_utf8_lossy(&o.stderr).trim().to_string(),
                            String::from_utf8_lossy(&o.stdout).trim().to_string(),
                        )
                    })
                    .unwrap_or_default();
                let duration_ms = t0.elapsed().as_millis();
                if status.success() {
                    return AcResult {
                        id: String::new(),
                        status: "pass".to_string(),
                        exit: Some(0),
                        duration_ms,
                        stderr_excerpt: String::new(),
                    };
                }
                let combined: String = [stderr, stdout]
                    .into_iter()
                    .filter(|s| !s.is_empty())
                    .collect::<Vec<_>>()
                    .join(" ")
                    .chars()
                    .take(100)
                    .collect();
                return AcResult {
                    id: String::new(),
                    status: "fail".to_string(),
                    exit: Some(status.code().map(i64::from).unwrap_or(1)),
                    duration_ms,
                    stderr_excerpt: combined,
                };
            }
            Ok(None) => {
                if Instant::now() >= deadline {
                    let _ = child.kill();
                    let _ = child.wait();
                    return AcResult {
                        id: String::new(),
                        status: "skip".to_string(),
                        exit: None,
                        duration_ms: t0.elapsed().as_millis(),
                        stderr_excerpt: format!("timeout after {}ms", AC_TIMEOUT_SECS * 1000),
                    };
                }
                std::thread::sleep(std::time::Duration::from_millis(50));
            }
            Err(_) => {
                return AcResult {
                    id: String::new(),
                    status: "skip".to_string(),
                    exit: None,
                    duration_ms: t0.elapsed().as_millis(),
                    stderr_excerpt: "wait failed".to_string(),
                };
            }
        }
    }
}

/// Emit the `qa.result` harness event.
fn emit_qa_event(cwd: &Path, spec: &str, overall: &str, criteria: &[Value]) {
    let store = JsonlEventStore::for_project(cwd);
    let ev = HarnessEvent {
        v: SCHEMA_VERSION,
        ts: now_iso8601(),
        session_id: session_id(),
        wave: 0,
        actor: Actor {
            kind: ActorKind::Cli,
            id: Some("qa-run".to_string()),
            actor_type: None,
        },
        event: "qa.result".to_string(),
        payload: json!({ "spec": spec, "overall": overall, "criteria": criteria }),
        spec: Some(spec.to_string()),
    };
    let _ = store.append(&ev);
}

/// Emit the `qa` metric (fail-silent).
fn emit_qa_metric(cwd: &Path, spec: &str, overall: &str, criteria: &[AcResult]) {
    let (mut pass, mut fail, mut skip) = (0, 0, 0);
    for c in criteria {
        match c.status.as_str() {
            "pass" => pass += 1,
            "fail" => fail += 1,
            "skip" => skip += 1,
            _ => {}
        }
    }
    let line = MetricLine::new(now_iso8601(), "qa").note(overall).extras(json!({
        "spec": spec,
        "overall": overall,
        "passCount": pass,
        "failCount": fail,
        "skipCount": skip,
        "category": "verification",
    }));
    let _ = emit_metric(cwd, &line);
}

/// The criteria array, as the JSON payload shape.
fn criteria_json(criteria: &[AcResult]) -> Vec<Value> {
    criteria
        .iter()
        .map(|c| {
            json!({
                "id": c.id,
                "status": c.status,
                "exit": c.exit,
                "duration_ms": c.duration_ms,
                "stderr_excerpt": c.stderr_excerpt,
            })
        })
        .collect()
}

/// Write the JSON sidecar at `.claude/.qa-reports/{spec}.json`.
fn write_sidecar(cwd: &Path, spec: &str, payload: &Value) {
    let dir = cwd.join(".claude").join(".qa-reports");
    if std::fs::create_dir_all(&dir).is_err() {
        return;
    }
    if let Ok(text) = serde_json::to_string_pretty(payload) {
        let _ = std::fs::write(dir.join(format!("{spec}.json")), text);
    }
}

/// Write the standalone HTML report at `.claude/.qa-reports/{spec}.html`.
fn write_html_report(cwd: &Path, spec: &str, overall: &str, criteria: &[AcResult]) -> Option<PathBuf> {
    let dir = cwd.join(".claude").join(".qa-reports");
    std::fs::create_dir_all(&dir).ok()?;
    let mut report = Report::new(
        format!("QA Report — {spec}"),
        format!("overall: {overall} · {} criteria", criteria.len()),
    );
    let rows: Vec<Vec<String>> = criteria
        .iter()
        .map(|c| {
            vec![
                c.id.clone(),
                c.status.to_uppercase(),
                c.exit.map(|e| e.to_string()).unwrap_or_else(|| "n/a".to_string()),
                format!("{:.1}s", c.duration_ms as f64 / 1000.0),
                c.stderr_excerpt.chars().take(80).collect(),
            ]
        })
        .collect();
    report.section(
        "Acceptance Criteria",
        &table(&["ID", "Status", "Exit", "Duration", "stderr"], &rows),
    );
    let path = dir.join(format!("{spec}.html"));
    std::fs::write(&path, report.render()).ok()?;
    Some(path)
}

/// Result of a QA run — `overall` plus the criteria.
struct QaResult {
    overall: String,
    criteria: Vec<AcResult>,
}

/// Run QA for `spec` under `cwd`. Always emits the event + metric.
fn run_qa(cwd: &Path, spec: &str) -> QaResult {
    let Some(spec_file) = find_spec_file(cwd, spec) else {
        eprintln!("[qa-run] Spec file not found for \"{spec}\"");
        emit_qa_metric(cwd, spec, "skip", &[]);
        return QaResult { overall: "skip".to_string(), criteria: Vec::new() };
    };
    let markdown = match std::fs::read_to_string(&spec_file) {
        Ok(m) => m,
        Err(err) => {
            eprintln!("[qa-run] Cannot read spec file: {err}");
            emit_qa_metric(cwd, spec, "skip", &[]);
            return QaResult { overall: "skip".to_string(), criteria: Vec::new() };
        }
    };
    let Some(section) = extract_ac_section(&markdown) else {
        eprintln!("[qa-run] WARN: No \"Acceptance Criteria\" section found in spec");
        emit_qa_event(cwd, spec, "skip", &[]);
        emit_qa_metric(cwd, spec, "skip", &[]);
        return QaResult { overall: "skip".to_string(), criteria: Vec::new() };
    };
    let items = parse_ac_items(&section);
    if items.is_empty() {
        eprintln!("[qa-run] WARN: Acceptance Criteria section found but no parseable AC items");
        emit_qa_event(cwd, spec, "skip", &[]);
        emit_qa_metric(cwd, spec, "skip", &[]);
        return QaResult { overall: "skip".to_string(), criteria: Vec::new() };
    }

    let mut criteria = Vec::new();
    let (mut fail_count, mut skip_count) = (0usize, 0usize);
    for item in &items {
        let mut res = run_ac_command(&item.command, cwd);
        res.id = item.id.clone();
        if res.status == "fail" {
            fail_count += 1;
        } else if res.status == "skip" {
            skip_count += 1;
        }
        criteria.push(res);
    }
    let overall = if fail_count > 0 {
        "fail"
    } else if skip_count == items.len() {
        "skip"
    } else {
        "pass"
    };

    let cjson = criteria_json(&criteria);
    let payload = json!({ "spec": spec, "overall": overall, "criteria": cjson });
    emit_qa_event(cwd, spec, overall, &cjson);
    emit_qa_metric(cwd, spec, overall, &criteria);
    write_sidecar(cwd, spec, &payload);

    QaResult { overall: overall.to_string(), criteria }
}

/// Dispatch `mustard-rt run qa-run`.
pub fn run(spec: &str, format: &str) {
    let cwd = std::env::current_dir()
        .ok()
        .or_else(|| Some(PathBuf::from(project_dir())))
        .unwrap_or_else(|| PathBuf::from("."));

    let result = run_qa(&cwd, spec);
    let cjson = criteria_json(&result.criteria);

    if format == "html" {
        match write_html_report(&cwd, spec, &result.overall, &result.criteria) {
            Some(path) => eprintln!("[qa-run] HTML report: {}", path.display()),
            None => eprintln!("[qa-run] WARN: could not write HTML report"),
        }
    }

    // JSON is always emitted on stdout (the pipeline-consumed contract).
    let out = json!({
        "event": "qa.result",
        "payload": { "spec": spec, "overall": result.overall, "criteria": cjson },
    });
    println!("{}", serde_json::to_string_pretty(&out).unwrap_or_else(|_| "{}".to_string()));

    if result.overall == "fail" {
        std::process::exit(1);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn parses_ac_lines_with_and_without_backticks() {
        let a = parse_ac_line("- [ ] AC-1: builds clean — Command: `cargo build`").unwrap();
        assert_eq!(a.id, "AC-1");
        assert_eq!(a.command, "cargo build");
        let b = parse_ac_line("- [x] AC-2: tests pass - Command: cargo test").unwrap();
        assert_eq!(b.id, "AC-2");
        assert_eq!(b.command, "cargo test");
        assert!(parse_ac_line("- just a bullet").is_none());
    }

    #[test]
    fn extracts_ac_section_body() {
        let md = "# Spec\n\n## Acceptance Criteria\n- [ ] AC-1: x — Command: `true`\n\n## Files\n- a.rs\n";
        let section = extract_ac_section(md).unwrap();
        assert!(section.contains("AC-1"));
        assert!(!section.contains("Files"));
    }

    #[test]
    fn skips_when_spec_missing() {
        let dir = tempdir().unwrap();
        let r = run_qa(dir.path(), "ghost");
        assert_eq!(r.overall, "skip");
    }

    /// An AC-style command with quotes AND parentheses must survive intact to
    /// the shell. Under the old `cmd.arg("/C").arg(command)` path, `std`'s
    /// `CommandLineToArgvW`-style quoting corrupts the line (`node` sees a
    /// split string → "Unterminated string constant"); the `raw_arg`-based
    /// `build_shell_command` passes it verbatim, so this exits 0.
    #[cfg(windows)]
    #[test]
    fn ac_command_with_quotes_and_parens_runs_verbatim() {
        let dir = tempdir().unwrap();
        // node one-liner: a regex test inside parentheses, double-quoted -e arg.
        let cmd = r#"node -e "process.exit(/^(foo|bar)$/.test('bar') ? 0 : 1)""#;
        let res = run_ac_command(cmd, dir.path());
        assert_eq!(
            res.status, "pass",
            "quoted+parenthesized AC command must run verbatim (exit {:?}, stderr: {})",
            res.exit, res.stderr_excerpt
        );
        assert_eq!(res.exit, Some(0));
    }

    /// A `cmd.exe`-native command echoing a parenthesized, quoted string — the
    /// simplest case proving the outer quote pair is stripped and the inner
    /// `()` reach the program unmangled.
    #[cfg(windows)]
    #[test]
    fn ac_command_echoes_parenthesized_string() {
        let dir = tempdir().unwrap();
        let cmd = r#"node -e "console.log('(ok)')""#;
        let res = run_ac_command(cmd, dir.path());
        assert_eq!(res.status, "pass", "stderr: {}", res.stderr_excerpt);
        assert_eq!(res.exit, Some(0));
    }

    #[test]
    fn html_report_is_standalone() {
        let dir = tempdir().unwrap();
        let criteria = vec![AcResult {
            id: "AC-1".into(),
            status: "pass".into(),
            exit: Some(0),
            duration_ms: 12,
            stderr_excerpt: String::new(),
        }];
        let path = write_html_report(dir.path(), "demo", "pass", &criteria).unwrap();
        let html = std::fs::read_to_string(path).unwrap();
        assert!(html.starts_with("<!doctype html>"));
        assert!(html.contains("<style>"));
        assert!(!html.contains("href=") && !html.contains("src="));
        assert!(html.contains("AC-1"));
    }
}
